From 9d4b7117e69d48bee65ae41f92c470af7d7708d5 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 28 Oct 2024 12:48:39 +0100
Subject: [PATCH] Revert "media: stm32: dcmipp: avoid calling s_stream if state
 already correct"

This reverts commit 74e563ab7038a47d30f6273fefb5cb0215b3f130.

Remove this method as it doesn't fix entirely the problem of multiple calls
to the v4l2_s_stream and doesn't allow to have correct concurrent access
handling to protect cases when multiple pipes are being used in parallel.

Change-Id: Ibd332754d56f13c5dc3e87db7e965bfef50f2d54
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/421287
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-bytecap.c    |  2 +-
 .../st/stm32/stm32-dcmipp/dcmipp-byteproc.c   |  4 +-
 .../st/stm32/stm32-dcmipp/dcmipp-common.h     | 39 -------------------
 .../st/stm32/stm32-dcmipp/dcmipp-input.c      |  4 +-
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        |  2 +-
 .../st/stm32/stm32-dcmipp/dcmipp-pixelcap.c   |  2 +-
 .../st/stm32/stm32-dcmipp/dcmipp-pixelproc.c  |  4 +-
 .../st/stm32/stm32-dcmipp/dcmipp-statcap.c    |  2 +-
 8 files changed, 10 insertions(+), 49 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
index a878ea811e10..4279c7b4a827 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-bytecap.c
@@ -375,7 +375,7 @@ static int dcmipp_pipeline_s_stream(struct dcmipp_bytecap_device *vcap,
 	if (ret < 0)
 		return ret;
 
-	ret = dcmipp_s_stream_helper(vcap->s_subdev, state);
+	ret = v4l2_subdev_call(vcap->s_subdev, video, s_stream, state);
 	if (ret < 0) {
 		dev_err(vcap->dev, "failed to %s streaming (%d)\n",
 			state ? "start" : "stop", ret);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c
index 73829b3ee881..cfaf96ddf49c 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c
@@ -604,7 +604,7 @@ static int dcmipp_byteproc_s_stream(struct v4l2_subdev *sd, int enable)
 		if (ret)
 			return ret;
 
-		ret = dcmipp_s_stream_helper(s_subdev, enable);
+		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0) {
 			dev_err(byteproc->dev,
 				"failed to start source subdev streaming (%d)\n",
@@ -612,7 +612,7 @@ static int dcmipp_byteproc_s_stream(struct v4l2_subdev *sd, int enable)
 			return ret;
 		}
 	} else {
-		ret = dcmipp_s_stream_helper(s_subdev, enable);
+		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0) {
 			dev_err(byteproc->dev,
 				"failed to stop source subdev streaming (%d)\n",
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h
index 00feca94d5cc..a779c4a17e5d 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h
@@ -248,45 +248,6 @@ static inline void __reg_clear(struct device *dev, void __iomem *base, u32 reg,
 	__reg_write(dev, base, reg, readl_relaxed(base + reg) & ~mask);
 }
 
-/*
- * Helper function to avoid calling s_stream if v4l2_subdev streaming state
- * is already correct
- * This helper is here temporarily while waiting to move to proper
- * v4l2_subdev_stream_ functions
- * The move isn't yet done since this would require a much larger rework such
- * as implementing pad->enable_streams ops for subdev having multiple source
- * pads (such as dcmipp_input).  Moreover the stream API is still considered
- * as experimental as of linux v6.6
- *
- * This function reuse the enabled_streams counter within v4l2-subdev to count
- * the number of users of a subdevice.  Below we need to use a local variable
- * and not directly enabled_streams otherwise we would have v4l2_subdev_call
- * complain that the value are already matching when enabling and disable the
- * subdev.
- */
-static inline int dcmipp_s_stream_helper(struct v4l2_subdev *sd, int state)
-{
-	u64 count = sd->enabled_streams;
-	int ret;
-
-	/*
-	 * If enabling, don't do anything if enabled_streams was already > 0
-	 * if disabling, don't do anything if decreamented enabled_streams is still > 0
-	 */
-	if ((state && count++) || (!state && --count)) {
-		sd->enabled_streams = count;
-		return 0;
-	}
-
-	ret = v4l2_subdev_call(sd, video, s_stream, state);
-	if (ret)
-		return ret;
-
-	sd->enabled_streams = count;
-
-	return 0;
-}
-
 /* DCMIPP subdev init / release entry points */
 struct dcmipp_ent_device *dcmipp_tpg_ent_init(const char *entity_name,
 					      struct dcmipp_device *dcmipp);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
index 3a1157f30db1..d2efac5aa64c 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
@@ -494,7 +494,7 @@ static int dcmipp_inp_s_stream(struct v4l2_subdev *sd, int enable)
 		if (ret)
 			return ret;
 
-		ret = dcmipp_s_stream_helper(s_subdev, enable);
+		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0) {
 			dev_err(inp->dev,
 				"failed to start source subdev streaming (%d)\n",
@@ -502,7 +502,7 @@ static int dcmipp_inp_s_stream(struct v4l2_subdev *sd, int enable)
 			return ret;
 		}
 	} else {
-		ret = dcmipp_s_stream_helper(s_subdev, enable);
+		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0) {
 			dev_err(inp->dev,
 				"failed to stop source subdev streaming (%d)\n",
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
index 68b3577544d0..03160080d5e5 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -599,7 +599,7 @@ static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 			return 0;
 	}
 
-	ret = dcmipp_s_stream_helper(s_subdev, enable);
+	ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 	if (ret < 0) {
 		dev_err(isp->dev,
 			"failed to start source subdev streaming (%d)\n", ret);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
index c69bf06ff10e..7bb2938f6505 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
@@ -495,7 +495,7 @@ static int dcmipp_pipeline_s_stream(struct dcmipp_pixelcap_device *vcap,
 	if (ret < 0)
 		return ret;
 
-	ret = dcmipp_s_stream_helper(vcap->s_subdev, state);
+	ret = v4l2_subdev_call(vcap->s_subdev, video, s_stream, state);
 	if (ret < 0) {
 		dev_err(vcap->dev, "failed to %s streaming (%d)\n",
 			state ? "start" : "stop", ret);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
index 70a05a91ae45..e9fa98e1db24 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
@@ -1064,7 +1064,7 @@ static int dcmipp_pixelproc_s_stream(struct v4l2_subdev *sd, int enable)
 	s_subdev = media_entity_to_v4l2_subdev(pad->entity);
 
 	if (!enable) {
-		ret = dcmipp_s_stream_helper(s_subdev, enable);
+		ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 		if (ret < 0)
 			dev_err(pixelproc->dev,
 				"failed to stop source subdev streaming (%d)\n",
@@ -1110,7 +1110,7 @@ static int dcmipp_pixelproc_s_stream(struct v4l2_subdev *sd, int enable)
 		return ret;
 	}
 
-	ret = dcmipp_s_stream_helper(s_subdev, enable);
+	ret = v4l2_subdev_call(s_subdev, video, s_stream, enable);
 	if (ret < 0) {
 		dev_err(pixelproc->dev,
 			"failed to start source subdev streaming (%d)\n", ret);
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
index 50e52758ac26..4483193ed7de 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
@@ -202,7 +202,7 @@ static int dcmipp_pipeline_s_stream(struct dcmipp_statcap_device *vcap,
 		vcap->s_subdev = media_entity_to_v4l2_subdev(pad->entity);
 	}
 
-	ret = dcmipp_s_stream_helper(vcap->s_subdev, state);
+	ret = v4l2_subdev_call(vcap->s_subdev, video, s_stream, state);
 	if (ret < 0) {
 		dev_err(vcap->dev, "failed to %s streaming (%d)\n",
 			state ? "start" : "stop", ret);
-- 
2.47.2

